moleculegen/evaluation/metric.py

```diff
@@ -88,6 +110,303 @@ def reset(self):
+
+        Parameters
+        ----------
+        index : int
+        metric : Metric
+
+        Raises
+        ------
+        TypeError
+            If `index` is out of range.
+            If `metric` is not of type `Metric`.
+        """
+        if not isinstance(metric, Metric):
+            raise TypeError(f'value must be of type Metric, not {type(metric)}')
+
+        try:
+            self._metrics.__setitem__(index, metric)
+        except IndexError as err:
+            err.args = ('metrics assignment index out of range',)
+            raise err
+
+    def __delitem__(self, index: int):
+        """Delete the metric in position `key`.
+
+        Parameters
+        ----------
+        index : int
+
+        Raises
+        ------
+        IndexError
+            If `index` is out of range.
+        """
+        try:
+            del self._metrics[index]
+        except IndexError as err:
+            err.args = ('metrics assignment index out of range',)
+            raise err
+
+    def __len__(self) -> int:
+        """Return the number of metrics in the sequence.
+
+        Returns
+        -------
+        int
+        """
+        return len(self._metrics)
+
+    def __contains__(self, metric: Metric) -> bool:
+        """Check if `metric` is in the sequence.
+
+        Parameters
+        ----------
+        metric : Metric
+
+        Returns
+        -------
+        bool
+        """
+        return metric in self._metrics
+
+    def __iter__(self) -> Iterator[Metric]:
+        """Iterate over the metrics in the sequence.
+
+        Yields
+        ------
+        metric : Metric
+        """
+        return iter(self._metrics)
+
+    def insert(self, index: int, metric: Metric):
+        """Insert `metric` in position `index`.
+
+        Parameters
+        ----------
+        index : int
+        metric : Metric
+
+        Raises
+        ------
+        TypeError
+            If `metric` is not of type `Metric`.
+        """
+        if index < len(self):
+            self.__setitem__(index, metric)
+        else:
+            if not isinstance(metric, Metric):
+                raise TypeError(f'value must be of type Metric, not {type(metric)}')
+
+            self._metrics.append(metric)
+
+    def pop(self, index: int = -1) -> Metric:
+        """Delete the metric in position `key`.
+
+        Parameters
+        ----------
+        index : int, default len(self)-1
+
+        Raises
+        ------
+        IndexError
+            If `index` is out of range.
+        """
+        try:
+            return self._metrics.pop(index)
+        except IndexError as err:
+            err.args = ('pop index out of range',)
+            raise
+
+    def add(self, metric, *metrics):
+        """Append metric(s) to the sequence.
+
+        Parameters
+        ----------
+        metric : Metric
+        metrics : tuple of Metric, default ()
+        """
+        self._metrics.extend(itertools.chain((metric,), metrics))
```

Comment:
For every metric in the chain, check if it is of type `moleculegen.evaluation.Metric`.